files/libs/xroad_libs.sh

```diff
@@ -41,189 +41,184 @@ function destroy_api_key () {
-    # signer_console exit code is always 0 :(
-    if [ "$(cat /tmp/signer_console.log | grep ERROR)" ]; then
-      cat /tmp/signer_console.log
-      exit 1
-    fi
-  else
-    log "sign certificate already imported"
+function add_timestamping_service() {
+  tpl='{
+  "name": "%s",
+  "url": "%s"
+}'
+  printf -v data "$tpl" "$PX_TSA_NAME" "$PX_TSA_URL"
+  request_api POST "/system/timestamping-services" "$data"
+  if [[ $api_response_status_code -ne 201 ]]; then
+    log "Warning, Could not add timestamping service. $api_response_body"
   fi
 }
 
-# subfunction to initialize token with expect
-# short password will cause Signer.TokenPinPolicyFailure
-# we're expecting {"event":"Initialize the software token",...}
-function expect_initialize_software_token() {
-  local pin=$1
-  export -f signer_console
-
-  expect -c "proc abort {} {
-               puts Aborted
-               exit 1
-             }
-             spawn /bin/bash -c { signer_console init-software-token };
-             expect 'PIN: ';
-             send \"$pin\r\";
-             expect 'retype PIN: ';
-             send \"$pin\r\";
-             expect {
-                \"token initialization failed\" abort
-                \"Signer.TokenPinPolicyFailure\" abort
-                \"\\\"event\\\":\\\"Initialize the software token\\\"\"
-             }
-             "
+function token_login() {
+  request_api PUT "/tokens/0/login" "{\"password\": \"$software_token_pin\"}"
 }
 
-function initialize_software_token() {
-  local pin=$1
-  export -f signer_console
+function generate_key_and_csr () {
+  local type=$1
+  if [ $type = auth ]; then
+    local key_usage_type=AUTHENTICATION
+  else 
+    local key_usage_type=SIGNING
+    local member_id=", \"member_id\": \"$PX_INSTANCE:$PX_MEMBER_CLASS:$PX_MEMBER_CODE\""
+  fi
 
-  if [ "$(signer_console list-tokens|grep OK)" ]; then
-    log "software token already initialized"
-  else
-    output=$(expect_initialize_software_token $pin)
-    expect_exit=$?
-    if [ $expect_exit -eq 0 ]; then
-      log "software token initialized"
+  tpl='{
+	"key_label": "%s-%s",
+	"csr_generate_request": {
+		"key_usage_type": "%s",
+		"ca_name": "%s",
+		"csr_format": "PEM",
+		"subject_field_values": {
+			"CN": "%s",
+			"SN": "%s"
+		}
+    %s
+	}
+}'
+  printf -v data "$tpl" "$type" "$PX_MEMBER_CODE" "$key_usage_type" "$PX_CA_NAME" "$PX_MEMBER_CODE" "$PX_MEMBER_CODE" "$member_id"
+
+  # assume token id: 0
+  request_api POST "/tokens/0/keys-with-csrs" "$data"
+  if [ $api_response_status_code = 200 ]; then
+    csr_data=($(echo "$api_response_body" | jq -r -c '.key.id,.csr_id'))
+    csr_path="/tmp/$type-$PX_MEMBER_CODE.csr"
+    # download csr to /tmp
+    $(curl -s -k -H "Authorization: X-Road-ApiKey token=${created_api_key[1]}" \
+    https://localhost:4000/api/v1/keys/${csr_data[0]}/csrs/${csr_data[1]}?csr_format=PEM -o $csr_path)
```

Comment:
saving result to /tmp folder - assuming these scripts are not reused somewhere else.

Reply:
if you wan't you can add the hash, but I would keep type and membercode so it would be easy to identify the correct CSR (though not sure if and when would this be needed).

Reply:
hm, but further, shouldn't you actually store the CSR in some persistent directory instead. I mean in the next step you are also checking if the CSR exists or not.

Reply:
hm, not sure actually. Ignore the above.

Reply:
it is possible not to store the csr and crt at all

Reply:
Let's store it in /etc/xroad/signer/ for debugging reasons.